### DISCUSSION TITLE
Offering limited Monte carlo processing?

### DISCUSSION THREAD

**Discussion by malcolm-dsider** [NatLabRockies/GEOPHIRES-X#91] [Ideas]
How about we set up a page (like https://jonathanpezzino.com/geothermal/geophires and https://jonathanpezzino.com/geothermal/hip-ra) that offers limited Monte Carlo simulations? The problem is that MC processes take a lot of resources, and while I don't know how much grunt the current server has, what if we limited the number to 1000 iterations (or whatever number makes sense)? So that people can do some testing for themselves before they decide to download the code and run it locally so they can do far more iterations? Or perhaps once we get a server with more grunt (say, via Gov Cloud), then we could just remove the limit.

It might actually be two pages - one to run HIP-RAX in MC mode, and one to run GEOPHIRES? And because GEOPHIRES is more compute intensive, we could set a lower limit?

**Comment by softwareengineerprogrammer** [comment]
@malcolm-dsider It would be great to offer MC processing similar to the existing GEOPHIRES & HIP-RA tools. There are a few layers of foundational work that need to be addressed for this to be possible, some of which I already anticipate doing as part of planned work for existing projects.

At this stage the immediate priority would be cleaning up code structure to create a well-defined/bounded interface for input/output to MC - as has already been done for GEOPHIRES with [`GeophiresXClient`](https://github.com/NREL/GEOPHIRES-X/blob/3d5d75477ba93df399f66646d433ff56a8e6bbe5/src/geophires_x_client/__init__.py#L14), and partially done in [`HipRaClient`](https://github.com/NREL/GEOPHIRES-X/blob/3d5d75477ba93df399f66646d433ff56a8e6bbe5/src/hip_ra/__init__.py#L30) with [additional work currently in progress](https://github.com/softwareengineerprogrammer/GEOPHIRES-X/blob/babf6edefa80353d8219b6e198befbcf66a8639f/src/hip_ra/__init__.py#L31-L45) as part of HIP-RA-X.

After that, there are a number of factors to consider from a service architecture and product standpoint, including - but not limited to - compute resources and number of allowed iterations. Given the high compute resource demand, it is almost certain that access would need to be gated by user registration to prevent abuse and cost overrun. On the whole, implementation would be a multi-part project requiring planning and investment extending beyond an incremental addition to the existing mechanisms, but definitely doable with the right resources and prioritization.

**Reply by softwareengineerprogrammer** [reply]
This work has been done and Monte Carlo has been available in the web interface since September 2024 at **https://gtp.scientificwebservices.com/monte-carlo/**. Users may run a limited number of iterations at no charge, or up to 2000 GEOPHIRES and 3000 HIP-RA iterations with SWS GTP API Access.
